ath79: support for TP-Link EAP225-Wall v2

svanheule opened the pull request:
TP-Link EAP225-Wall v2 is an AC1200 (802.11ac Wave-2) wall plate access point. UART access and debricking require fine soldering. 

The device was kindly provided for porting by Stijn Segers.

I just managed to destroy the device, but the patches were tested earlier.

svanheule committed:
ath79: support for TP-Link EAP225-Wall v2

TP-Link EAP225-Wall v2 is an AC1200 (802.11ac Wave-2) wall plate access
point. UART access and debricking require fine soldering.

The device was kindly provided for porting by Stijn Segers.

Device specifications:
* SoC: QCA9561 @ 775MHz
* RAM: 128MiB DDR2
* Flash: 16MiB SPI-NOR (GD25Q127CSIG)
* Wireless 2.4GHz (SoC): b/g/n, 2x2
* Wireless 5Ghz (QCA9886): a/n/ac, 2x2 MU-MIMO
* Ethernet (SoC): 4× 100Mbps
  * Eth0 (back): 802.3af/at PoE in
  * Eth1, Eth2 (bottom)
  * Eth3 (bottom): PoE out (can be toggled by GPIO)
* One status LED
* Two buttons (both work as failsafe)
  * LED button, implemented as KEY_BRIGHTNESS_TOGGLE
  * Reset button

Flashing instructions, requires recent firmware (tested on 1.20.0):
* ssh into target device and run `cliclientd stopcs`
* Upgrade with factory image via web interface

Debricking:
* Serial port can be soldered on PCB J4 (1: TXD, 2: RXD, 3: GND, 4: VCC)
    * Bridge unpopulated resistors R162 (TXD) and R165 (RXD)
      Do NOT bridge R164
    * Use 3.3V, 115200 baud, 8n1
* Interrupt bootloader by holding CTRL+B during boot
* tftp initramfs to flash via sysupgrade or LuCI web interface

MAC addresses:
MAC address (as on device label) is stored in device info partition at
an offset of 8 bytes. ath9k device has same address as ethernet, ath10k
uses address incremented by 1.
From OEM ifconfig:
    br0       Link encap:Ethernet  HWaddr 50:...:04
    eth0      Link encap:Ethernet  HWaddr 50:...:04
    wifi0     Link encap:UNSPEC  HWaddr 50-...-04-...
    wifi1     Link encap:UNSPEC  HWaddr 50-...-05-...

Signed-off-by: Sander Vanheule <sander@svanheule.net>

adschm commented on target/linux/ath79/generic/base-files/etc/board.d/03_gpio_switches:
```diff
@@ -73,6 +73,9 @@ tplink,wbs510-v1|\
 	;;
+tplink,eap225-wall-v2)
+	ucidef_add_gpio_switch "poe_passthrough" "PoE Passthrough" "tp-link:poe-passthrough:enable"
+	;;
```
gpio-export and 03_gpio_switches are _alternatives_

Either keep your gpio-export implementation and drop this entry in 03_gpio_switches, or
drop gpio-export and put `ucidef_add_gpio_switch "poe_passthrough" "PoE Passthrough" "20"` here (so, you can effectively merge with the cpe/wbs devices).

Which one you choose is essentially up to you.

svanheule replied:
If I only add a named export in the DTS, would that still create an entry by default in the system config? The switch is off by default, but I would like the UCI config to be present already so it is more discoverable to users.

adschm replied:
I'm not 100 % sure, but I think gpio-export can be toggled by the user, but not via uci:

`echo 1 > /sys/class/gpio/gpio20/value`

gpio-hog can't be toggled at all, and if you want uci config, you need to go with 03_gpio_switches.

svanheule replied:
Going through the init code, named gpio pins in a gpio_switch config are [handled differently](https://github.com/openwrt/openwrt/blob/f1bc66c7653e7603b4621f5ea14be129a2bd5e4f/package/base-files/files/etc/init.d/gpio_switch#L46). For a numerical gpio_pin value, the gpio is exported, configured and has its value set correctly. A named gpio pin is assumed to be set up already, and only the saved value is restored. This was introduced in commit c5bf9a8ced6567c6bd9da63c6ff4c4f2f87207fd.

By using 03_gpio_switches, I can ensure /etc/config/system is preseeded with the appropriate gpio_switch section. But if you feel this is something that should be left to the user, I can drop this part and only use the DTS export.

svanheule replied:
Since it looks like 03_gpio_switches is never used for named gpio's, I'll just stick to gpio-export. I would still like the preseeded config, but this is probably something that needs be solved in a more generic way with another patch.

adschm replied:
> Since it looks like 03_gpio_switches is never used for named gpio's,

Just look at how "new" the referenced commit is (and that I didn't know about it as well). But actually I'm not really sure whether I like this new setup option created there; so I'm happy with pure gpio-export as you did now.

Keep in mind that both 03_gpio_switches and gpio-export are actually just intermediate workarounds that only exist because nobody is interested in doing a proper implementation (gpio-export is naked upstream, and 03_gpio_switches is pure user-space). The available upstream solution (gpio-hog) however does not support user interaction, though ...

adschm commented on target/linux/ath79/image/generic-tp-link.mk:
```diff
@@ -364,16 +364,27 @@ TARGET_DEVICES += tplink_cpe610-v2
+  IMAGE_SIZE := 15104k
+  DEVICE_MODEL := EAP225-Wall
+  DEVICE_VARIANT := v2
+  TPLINK_BOARD_ID := EAP225-WALL-V2
```
Consider putting TPLINK_BOARD_ID after DEVICE_PACKAGES to match the order of subsequent devices.

svanheule replied:
Fixed.

adschm commented on tools/firmware-utils/src/tplink-safeloader.c:
```diff
@@ -1298,6 +1298,35 @@ static struct device_info boards[] = {
+			{"product-info", 0x31100, 0x00400},
+			{"soft-version", 0x32000, 0x00100},
+			{"firmware", 0x40000, 0xd80000},
+			{"user-config", 0xdc0000, 0x30000},
```
This doesn't match the partitions in DTS. Here, you have a gap from 0xe00000 to 0xf30000, and in partitions you have the config partition at a different location?

svanheule replied:
I'm afraid this is one of those TP-Link screw-ups. I've put the config partition in the DTS at the location determined by the original bootargs. I could correct the partition table here, but this is the way it's described in stock.

From FW v1.20.0:
```
partition fs-uboot base 0x00000 size 0x20000
partition partition-table base 0x20000 size 0x02000
partition default-mac base 0x30000 size 0x01000
partition support-list base 0x31000 size 0x00100
partition product-info base 0x31100 size 0x00400
partition soft-version base 0x32000 size 0x00100
partition os-image base 0x40000 size 0x180000
partition file-system base 0x1c0000 size 0xc00000
partition user-config base 0xdc0000 size 0x30000
partition mutil-log base 0xf30000 size 0x80000
partition oops base 0xfb0000 size 0x40000
partition radio base 0xff0000 size 0x10000
```
Bootargs:
```
mtdparts=spi0.0:
    128k(u-boot),
    64k(pation-table),
    64k(product-info),
    1536k(kernel),
    13568k(rootfs),
    192k(config),
    512k(mutil-log),
    256k(oops),
    64k(ART)
```

Boot log:
```
[    0.470000] Creating 9 MTD partitions on "spi0.0":
[    0.480000] 0x000000000000-0x000000020000 : "u-boot"
[    0.490000] 0x000000020000-0x000000030000 : "pation-table"
[    0.490000] 0x000000030000-0x000000040000 : "product-info"
[    0.500000] 0x000000040000-0x0000001c0000 : "kernel"
[    0.500000] 0x0000001c0000-0x000000f00000 : "rootfs"
[    0.510000] mtd: partition "rootfs" set to be root filesystem
[    0.520000] 0x000000f00000-0x000000f30000 : "config"
[    0.520000] 0x000000f30000-0x000000fb0000 : "mutil-log"
[    0.530000] 0x000000fb0000-0x000000ff0000 : "oops"
[    0.540000] 0x000000ff0000-0x000001000000 : "ART"
```

adschm replied:
How nice...

Is there actually config data in the marked partition (if this can be determined at all)?

I don't have a strong opinion here, but having inconsistent safeloader and partitions seems undesirable to me. I assume the config partition still is recreated anyway if overwritten like with older devices?

adschm replied:
Yes, on a 16M device I'd also prefer to be safe and keep the partition as suggested.

adschm replied:
>  I had dumped the the NAND

I thought this is SPI-NOR?

svanheule replied:
Sorry, my mistake. It's SPI-NOR.

svanheule commented:
I've modified the DTS partition layout to match the safeloader layout. The DTS has a comment to explain to future code archaeologists why there's a hole in the list.

The 03_gpio_switches patch was also dropped to match the implementation of other devices.

adschm commented:
Thanks, will merge during next hour.

svanheule commented:
Great! Thanks for the feedback, once again.
